File: Chapter-1/binarySearch.ts
interface SearchProps<T> {
  arr: T[];
  target: T;
}

function binarySearch<T>({ arr, target }: SearchProps<T>): number {
    let left = 0;
    let right = arr.length - 1;
    let index = 0

    while (left <= right) {
        index = Math.floor((left + right) / 2); 
        if (arr[index] === target) {
            return index;
        } else if (arr[index] < target) {
            left = index + 1;
        } else {
            right = index - 1;
        } 
    }          

    return -1;
}

const numbers = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
const targetNumber = 29;
const resultIndex = binarySearch({ arr: numbers, target: targetNumber });
const foundMessage = resultIndex !== -1 ? `at found index: ${resultIndex}` : "not found";
const stringTemplate = `Target number ${targetNumber} ${foundMessage}`;
console.log(stringTemplate);  

const alphabet = ["a" , "b", "c", "d", "e"];
const targetSrt = "c";
const resultIndexCh = binarySearch({ arr: alphabet, target: targetSrt });
console.log(`Target string ${targetSrt} found at index: ${resultIndexCh}`);

File: Chapter-1/binarySearch.py
def binary_search(arr, target):
    """
    Perform binary search on a sorted array to find the index of the target value.

    Parameters:
    arr (list): A list of sorted elements.
    target: The element to search for in the list.

    Returns:
    int: The index of the target element if found, otherwise -1.
    """
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2

        # Check if target is present at mid
        if arr[mid] == target:
            return mid
        # If target is greater, ignore left half
        elif arr[mid] < target:
            left = mid + 1
        # If target is smaller, ignore right half
        else:
            right = mid - 1

    # Target was not found in the array
    return -1